mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoCRUDController.java

```diff
@@ -532,60 +532,78 @@ private void createUpdateEntityInfoIndexes(EntityInfo ei) {
-                            deleteIndexes.remove(existingIndex);
+            // This is how index creation/modification works:
+            //  - The _id index will remain untouched.
+            //  - If there is an index with name X in metadata, find the same named index, and compare
+            //    its fields/flags. If different, drop and recreate. Don't do this for _id index.
+            //  - If there is an index with null name in metadata, see if there is an index with same
+            //    fields and flags. If so, no change. Otherwise, create index. Don't do this for _id index
+            List<Index> createIndexes=new ArrayList<>();
+            List<DBObject> dropIndexes=new ArrayList<>();
+            for(Index index:indexes.getIndexes()) {
+                if(!isIdIndex(index)) {
+                    if(index.getName()!=null&&index.getName().trim().length()>0) {
+                        LOGGER.debug("Processing index {}",index.getName());
+                        DBObject found=null;
+                        for(DBObject existingIndex:existingIndexes) {
+                            if(index.getName().equals(existingIndex.get("name"))) {
+                                found=existingIndex;
+                                break;
+                            }
+                        }
+                        if(found!=null) {
+                            if(indexFieldsMatch(index,found) &&
+                               indexOptionsMatch(index,found) ) {
+                                LOGGER.debug("{} already exists",index.getName());
+                            } else {
+                                LOGGER.debug("{} modified, dropping and recreating index",index.getName());
+                                dropIndexes.add(found);
+                                createIndexes.add(index);
+                            }
+                        } else {
+                            LOGGER.debug("{} not found, creating index",index.getName());
+                            createIndexes.add(index);
+                        }
+                    } else {
+                        LOGGER.debug("Processing index with fields {}",index.getFields());
+                        DBObject found=null;
+                        for(DBObject existingIndex:existingIndexes) {
+                            if(indexFieldsMatch(index,existingIndex)&&
+                               indexOptionsMatch(index,existingIndex)) {
+                                found=existingIndex;
+                                break;
+                            }
+                        }
+                        if(found!=null) {
+                            LOGGER.debug("An index with same keys found: {}",found);
+                        } else {
+                            LOGGER.debug("Creating index with fields {}",index.getFields());
+                            createIndexes.add(index);
```

Comment:
How will this function if there is an unnamed index with these fields / options?  Will there then be two indexes, one named and one unnamed?